src/lib/pubkey/xmss/xmss_privatekey.cpp

```diff
@@ -32,105 +34,201 @@ namespace Botan {
+         }
 
-   if(unused_leaf >= (1ull << XMSS_PublicKey::m_xmss_params.tree_height()))
-      {
-      throw Decoding_Error("XMSS private key leaf index out of bounds");
-      }
+      void set_unused_leaf_index(size_t idx)
+         {
+         if(idx >= (1ull << m_xmss_params.tree_height()))
+            {
+            throw Decoding_Error("XMSS private key leaf index out of bounds");
+            }
+         else
+            {
+            std::atomic<size_t>& index =
+               static_cast<std::atomic<size_t>&>(*recover_global_leaf_index());
+            size_t current = 0;
+
+            do
+               {
+               current = index.load();
+               if(current > idx)
+                  { return; }
+               }
+            while(!index.compare_exchange_strong(current, idx));
+            }
+         }
+
+      size_t reserve_unused_leaf_index()
+         {
+         size_t idx = (static_cast<std::atomic<size_t>&>(
+                        *recover_global_leaf_index())).fetch_add(1);
+         if(idx >= (1ull << m_xmss_params.tree_height()))
+            {
+            throw Decoding_Error("XMSS private key, one time signatures exhaused");
+            }
+         return idx;
+         }
 
-   begin = end;
-   end = begin + XMSS_PublicKey::m_xmss_params.element_size();
-   m_prf.clear();
-   m_prf.reserve(XMSS_PublicKey::m_xmss_params.element_size());
-   std::copy(begin, end, std::back_inserter(m_prf));
+      size_t unused_leaf_index() const
+         {
+         return *recover_global_leaf_index();
+         }
 
-   begin = end;
-   end = begin + m_wots_params.element_size();
-   m_wots_priv_key.set_private_seed(secure_vector<uint8_t>(begin, end));
-   set_unused_leaf_index(static_cast<size_t>(unused_leaf));
-   }
+   private:
+      const XMSS_Parameters& m_xmss_params;
+      const XMSS_WOTS_Parameters& m_wots_params;
+
+      XMSS_WOTS_PrivateKey m_wots_priv_key;
+      XMSS_Hash m_hash;
+      secure_vector<uint8_t> m_prf;
+      XMSS_Index_Registry& m_index_reg;
+   };
+
+
+XMSS_PrivateKey::XMSS_PrivateKey(std::span<const uint8_t> key_bits)
+   : XMSS_PublicKey(key_bits)
+   , m_private(std::make_shared<XMSS_PrivateKey_Internal>(
+         m_xmss_params, m_wots_params, m_public_seed, key_bits)) {}
 
 XMSS_PrivateKey::XMSS_PrivateKey(
    XMSS_Parameters::xmss_algorithm_t xmss_algo_id,
    RandomNumberGenerator& rng)
-   : XMSS_PublicKey(xmss_algo_id, rng),
-     m_wots_priv_key(XMSS_PublicKey::m_xmss_params.ots_oid(),
-                     public_seed(),
-                     rng),
-     m_hash(xmss_hash_function()),
-     m_prf(rng.random_vec(XMSS_PublicKey::m_xmss_params.element_size())),
-     m_index_reg(XMSS_Index_Registry::get_instance())
+   : XMSS_PublicKey(xmss_algo_id, rng)
+   , m_private(std::make_shared<XMSS_PrivateKey_Internal>(
+      m_xmss_params, m_wots_params, m_public_seed,
+      rng.random_vec(m_xmss_params.element_size()),
+      rng.random_vec(m_xmss_params.element_size())))
```

Comment:
Re: CLI test failure: This reverses the order that secret seed and the PRF value are drawn from the RNG. The `cli_xmss_sign_tests` test verifies the hash of the private key given a fixed RNG-seed and therefore failed.

We could switch out the parameters in the `XMSS_PrivateKey_Internal` constructor to avoid changing the expecations of the test. I did that locally to make sure the test would turn green: ✔️
Though, I think, a sensible order of parameters trumps changing the test.

Reply:
OK I've given this some consideration and I think we should *not* change how the key is generated, and instead swap the order of the parameters.

In general, right now we do not guarantee that a particular RNG state will lead to a particular private key, if you re-run key generation. We do implicitly guarantee it for Kyber and Dilitihium, and probably will also for most of the other new PQC algorithms simply because of how the KATs are generated. This is largely a problem of underspecification for older algorithms.

I want to start moving in that direction by adding explicit tests (#3269) that check this behavior, without guaranteeing long term stability to end users (yet). That is, if we really decide to change how ECDSA keygen is done, we can do so, and just also update the tests. But eventually we would make such a guarantee, and the test data then becomes set in stone.

Reply:
Generally speaking: I'm fully in favor of this idea. Specifically for XMSS, though: Currently the key generation in Botan is not compliant to the suggestions in NIST SP.800-208 anyway. TL;DR: They introduce an additional PRF dedicated to key-generation to protect against a multi-target attack. Currently, Botan's implementation adheres to the recommendations in [RFC 8391](https://www.rfc-editor.org/rfc/rfc8391) where the specifics of key generation are not set in stone.

Long story short: When changing the key generation to the NIST spec, the relationship between RNG state and resulting private key will break for XMSS anyway. Note though, that the change in the key generation would not invalidate legacy private keys and/or signatures.

Reply:
OK given above situation, fine to proceed.